oak/server/rust/oak_runtime/src/channel.rs

```diff
@@ -0,0 +1,255 @@
+use std::prelude::v1::*;
+
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::vec::Vec;
+
+use core::sync::atomic::AtomicUsize;
+use core::sync::atomic::Ordering::SeqCst;
+
+use oak_abi::OakStatus;
+
+use crate::{Message, RuntimeRef};
+
+type Messages = VecDeque<Message>;
+
+/// The internal implementation of a channel representation backed by a `VecDeque<Message>`.
+/// Readers and writers to this channel must increment the reader/writer count. This is implemented
+/// for `ChannelWriter`/`ChannelReader`, which will increment/decrement readers/writers when
+/// cloning/dropping.
+struct Channel {
+    readers: AtomicUsize,
+    writers: AtomicUsize,
+    messages: RwLock<Messages>,
+}
+
+#[derive(Clone)]
+/// Reference to a `Channel` that is `Clone`able and can be passed across threads. Channels are
+/// multi-writer mult-reader.
+struct ChannelRef(Arc<Channel>);
+
+/// Writer end to a `Channel`.
+pub struct ChannelWriter(ChannelRef);
+/// Reader end to a `Channel`.
+pub struct ChannelReader(ChannelRef);
+
+/// A wrapper type to allow taking channel references without discriminating on direction. Used
+/// when adding `ChannelRef`s into `Message`s.
+#[derive(Clone)]
+pub enum ChannelEither {
+    Writer(ChannelWriter),
+    Reader(ChannelReader),
+}
+
+impl Channel {
+    /// Thread safe method that returns true when there is no longer at least one reader and one
+    /// writer.
+    fn is_orphan(&self) -> bool {
+        self.readers.load(SeqCst) == 0 || self.writers.load(SeqCst) == 0
+    }
+}
+
+/// Creates a new `ChannelWriter` and `ChannelReader` that reference the same underlying `Channel`.
+pub fn new() -> (ChannelWriter, ChannelReader) {
+    let chan = Arc::new(Channel {
+        readers: AtomicUsize::new(1),
+        writers: AtomicUsize::new(1),
+        messages: RwLock::new(VecDeque::new()),
+    });
+    let writer = ChannelWriter(ChannelRef::from_arc(chan.clone()));
+    let reader = ChannelReader(ChannelRef::from_arc(chan.clone()));
+    (writer, reader)
+}
+
+impl ChannelRef {
+    /// Internal method to take a thread-safe reference-counting pointer to a `Channel`.
+    fn from_arc(arc: Arc<Channel>) -> ChannelRef {
+        ChannelRef(arc)
+    }
+
+    /// Attempt to apply a function that takes a `&mut VecDeque<Message>` to the underlying message
+    /// store. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if the underlying channel has been
+    /// orphaned.
+    fn try_write<F, T>(&self, f: F) -> Result<T, OakStatus>
+    where
+        F: FnOnce(&mut Messages) -> T,
+    {
+        if self.0.is_orphan() {
```

Comment:
Is there a race condition here between checking `is_orphan` and actually enqueuing the message? i.e. could it become orphan right after this check? It seems to me that you would need to synchronize across `readers`, `writers` and `messages` with a single lock in order to avoid it, but I may be missing something.

Reply:
Even if we hold a lock across `readers`, `writers` and `messages`, the last reader can close after the write without reading the last message written to the queue. The end state is the same as the race condition between `is_orphan` and message enqueue- a message is left in the channel with no reader (importantly no violation of other properties).